### DISCUSSION TITLE
Cell Centre Data Outputting

### DISCUSSION THREAD

**Discussion by AlanZandie** [firemodels/fds#12659] [General]
Good day to you,

I have been involved with using certain optimisation techniques along with FDS to optimise a car park configuration, however, in doing so, I need the data fro velocity, visibility, temperature etc. at the centre of each cell. Is there a way that FDS can output that for me?

thank you

**Comment by rmcdermo** [comment]
Yes, it depends on which output you are using.  For DEVC, the data is automatically cell centered, it only depends on which I,J,K the DEVC position is located, unless you explicitly ask for interpolation.  For velocity, the primitive variables are staggered (face centered).  If you want this interpolated to a cell center, use `QUANTITY='CELL U'`, for u component, etc.

For SLCF (slice files) you can add `CELL_CENTERED=T`, generally.  For velocity, again, you need to look at `QUANTITY='CELL U'`.

**Comment by AlanZandie** [comment]
Thank you but what I meant was to imagine you have 300 thousand cells, I am looking for a way you could have the value of those features (T, Vel, Vis etc.) at each cell centre reported individually. For instance, something like this: 
Velocity predicted at:
Cell 1: XXX
Cell 2: YYY
.
.
.
you see I am searching for the raw data of these features at every cell. Not averaged over a region or at a particular spot, all the cells data is what I'm after.!

**Reply by rmcdermo** [reply]
The closest we have to dumping all the data is 3D slice files.  Search the user guide on `DT_SL3D`.  Beyond this, you would need to din into the source code and write your own output routines.

**Comment by TristanHehnen** [comment]
You could use the [fdsreader, a Python module ](https://firedynamics.github.io/LectureFireSimulation/content/tools/03_analysis/02_fdsreader.html)to read FDS output.

You can then, for example, define `SLCF` in FDS, with `QUANTITY='TEMPERATURE'` and `CELL_CENTERED = .TRUE.`.
with the fdsreader you can read the `SLCF` and you get a multidimensional numpy array, with the temperature values of each cells covered in the `SLCF` and for each time step, i.e. `NFRAMES` or `DT_SLCF`.
Same works with `BNDF` too.

**Comment by AlanZandie** [comment]
I'll try that @TristanHehnen 
Thank you, could you please keep this ticket open in case if I need more help, I'll pop in again.
Cheers

**Comment by TristanHehnen** [comment]
I believe it also works with 3d `SLCF`.

**Comment by AlanZandie** [comment]
Good day, 
I have been working with fdsreader to analyse the FDS outputs, while it works fine on the python installed on my laptop, it can't be recognised on the HPC linux I'm using. what I did was that I installed fdsreader using:
pip install git+https://github.com/FireDynamics/fdsreader.git
even though I exported it to PATH as well using: export PATH=$PATH:/path/to/fdsreader, I still get the error that fdsreader is not found. I understand it also needs incremental, typing-extensions and numpy, but all of them are installed and located at the same location as below:
/rds/homes/z/zandiem/.local/lib/python3.6/site-packages
FYI, I have attached a copy of the python script I'm using, and the error I get.  
could you please help me if you know what's the issue?. bear in mind, the error says incremental is not found, but I tried different combination to import fdsreader and its pre-requisits and i still get the same error for each.  
[Python code.txt](https://github.com/firemodels/fds/files/14744816/Python.code.txt)
[error.txt](https://github.com/firemodels/fds/files/14744824/error.txt)

**Reply by johodges** [reply]
Looks like something is wrong in your python environment. Your error file lists python 3.10 but the site package you listed is python 3.6. If you want to swap to pyfdstools (https://github.com/johodges/pyfdstools) I can provide some support, but I do not have a lot of experience with fdsreader.

**Reply by AlanZandie** [reply]
thank you for this, given that I've already written the code based on fdsreader, I might have to stick to it for now, but I'm looking for a way to resolve this by adding 'fdsreader' module to where python 3.10 is. can you help me with this issue? because I reckon this would work.

**Reply by johodges** [reply]
Try configuring a virtual env on the hpc, then install fdsreader to that environment. You can see an example of how to do that in the pyfdstools README, just replace the pip install pyfdstools with pip install fdsreader

**Reply by TristanHehnen** [reply]
I believe these virtual Python environments are a good idea in general. I tend to not use them and occasionally mess up my Python installation with different package dependencies, very annoying... :/

**Comment by TristanHehnen** [comment]
Sorry, I'm not quite sure I follow. It seems to me that you cloned the git repo itself, instead of installing the fdsreader.

Installing the fdsreader should simply be `pip install fdsreader`.
For importing the module you should only need `import fdsreader` in the script you are working with, and nothing else (the dependencies should be loaded from fdsreader automatically, if present).

Could you please try to only start Python and then only import the fdsreader? 


Then, the `error.txt` says that in the `access_script.py`, in line 50, it can't import the `incremental` module. 
Please do the same thing with `incremental`: start Python and only load this module. If it does not work, please check if it is installed and [in what version](https://stackoverflow.com/questions/10214827/find-which-version-of-package-is-installed-with-pip). To my understanding, it should be `incremental==22.10.0`, see the [requirements in the git repo](https://github.com/FireDynamics/fdsreader/blob/26921a244485b25a6b70d024359d40027592ed2b/requirements.txt#L49).


Otherwise, maybe you could also try to point to the path of the module manually, [like explained here](https://stackoverflow.com/questions/67631/how-can-i-import-a-module-dynamically-given-the-full-path):
```
import sys
# the mock-0.3.1 dir contains testcase.py, testutils.py & mock.py
sys.path.append('/foo/bar/mock-0.3.1')

from mock import Mock
```

Does this help?

**Reply by TristanHehnen** [reply]
With `pip install` there should also be no need to add it to the `PATH`, because it ends up in these `egg` directories, where Python automatically looks first for modules.

**Comment by AlanZandie** [comment]
Thank you everyone for your comments, I believe I took care of, the issue was that there were two versions of Python installed on our cluster, and it was searching through wrong module library. Also, given that I was not allowed to make changes on the cluster, I had to make my own repository and specify it using PATH, I did the same for PYTHONPATH, to look into the directory I saved my modules in, particularly fdsreader. 
Thank you,
